CMakePresets.json

```diff
@@ -0,0 +1,70 @@
+{
+    "version": 6,
+    "cmakeMinimumRequired": {
+        "major": 3,
+        "minor": 14,
+        "patch": 0
+      },
+    "configurePresets": [
+        {
+            "name": "common",
+            "hidden": true,
+            "cacheVariables": {
+                "FREECAD_USE_PYBIND11:BOOL": "ON"
+            }
+        },
+        {
+            "name": "debug",
+            "displayName": "Debug",
+            "description": "Default debug profile",
+            "binaryDir": "${sourceDir}/build/debug",
+            "inherits": [
+                "common"
+            ],
+            "cacheVariables": {
+                "CMAKE_BUILD_TYPE": "Debug"
+            }
+        },
+        {
+            "name": "release",
+            "displayName": "Release",
+            "description": "Default release profile",
+            "binaryDir": "${sourceDir}/build/release",
+            "inherits": [
+                "common"
+            ],
+            "cacheVariables": {
+                "CMAKE_BUILD_TYPE": "Release"
+            }
+        },
+        {
+            "name": "conda",
+            "hidden": true,
+            "cmakeExecutable": "${workspaceFolder}/.conda/freecad/bin/cmake",
+            "cacheVariables": {
+                "BUILD_WITH_CONDA:BOOL": "ON",
+                "FREECAD_USE_EXTERNAL_SMESH:BOOL": "ON",
+                "CMAKE_C_COMPILER": "${workspaceFolder}/.conda/freecad/bin/cc",
+                "CMAKE_CXX_COMPILER": "${workspaceFolder}/.conda/freecad/bin/c++"
```

Comment:
I think this works for linux but might not for macos and windows, is there a way to make these platform dependent?

Reply:
And that's a potential problem, that's why I called for testing. My thought is that `cc` and `c++` usually are symlinks to the correct compiler, hoping it works on all platforms.

Reply:
I'm not sure if this is even needed, at least if the conda environment has been activated. We need testing but I'm still struggling to get a grip on vscode and conda.

Reply:
Perhaps you're right, cmake can discover compilers by itself. We need an opinion from the expert @oursland

Reply:
haven't actually tested this vscode setup but when activating the environment with conda I found that CC and CXX where already appropriately set on windows and linux, I would expect the same bahvior in macos. So these variables should not be needed.

Reply:
Several items related to macOS development with Conda:

* Launching VS Code using the normal means results in `CC` and `CXX` being set to `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc` and `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++`.  These are not the correct compilers for building in a conda environment.
* Launching VS Code from a terminal after activating the conda environment will assign the `CC` and `CXX` variables to the conda compiler, however this method of launching VS Code is less than ideal.  The configuration is global for that instance, which may pose surprises for some.  When the instance is relaunched due to package upgrades and reboots, it will not persist the environment.  To address this, the script `.vscode/env.sh` was included and added as a pre-build task to ensure the environment was set correctly.
* On macOS with Conda, `${workspaceFolder}/.conda/freecad/bin/cc` and `${workspaceFolder}/.conda/freecad/bin/c++` do not exist.  The correct compilers are `${workspaceFolder}/.conda/freecad/bin/clang` and `${workspaceFolder}/.conda/freecad/bin/clang++` which was used in `.vscode/cmake-kits.json`
* CMake is now generating Makefiles instead of Ninja build files.  Restoring the Ninja generator to the configuration did not address the issue, so I am uncertain how to bring that back.
* CMake is performing a configure, but failing to build.  There's no output from a build, just an error -1.

I'll look into what is causing the last couple of issues, but overall I'd like again bring up the [position I stated 3 weeks ago](https://github.com/FreeCAD/FreeCAD/pull/9663#issuecomment-1563436021).

At this time the weekly builds and official releases are all generated via Conda.  To my knowledge, the only part of the build ecosystem that is not using Conda is the GitHub CI, which is using a [Dockerfile](https://github.com/FreeCAD/FreeCAD/blob/master/ci/Dockerfile) that doesn't reflect the current official build environment.  Moving toward a build system that prioritizes developer workstation-specific environment instead of standardized environments in source control may prove to be challenging w.r.t. developer onboarding and support.

Reply:
Thank you for the insight. 
From the first two points I understand automatic compiler discovery works IF the conda environment is activated first. The script `.vscode/env.sh` has been replaced with the python script `.vscode/ActivateCondaIfUsingConda.py`, which is run before CMake: configure and CMake: build. It should correctly activate the environment.
The clang compilers could get detected automaticaly IF conda is activated, correct? So we could omit hard-coding for those in the conda preset.

 > Adding "generator": "Ninja", to the conda preset restored the functionality 

Will fix.

> CMake is performing a configure, but failing to build. There's no output from a build, just an error -1

I've seen that happen when building a preset not yet configured

Reply:
> The script `.vscode/env.sh` has been replaced with the python script `.vscode/ActivateCondaIfUsingConda.py`, which is run before CMake: configure and CMake: build. It should correctly activate the environment.

I'll test to make sure this is working.  It could have struggled with the incorrect `CC` and `CXX`.

> > CMake is performing a configure, but failing to build. There's no output from a build, just an error -1
> 
> I've seen that happen when building a preset not yet configured

That could be what happened.  I'll retry.

Reply:
I recommend reading the steps I've written here: https://github.com/FreeCAD/DevelopersHandbook/pull/52 see if that fixes the build error

Reply:
I haven't been able to get `ActivateCondaIfUsingConda.py` to launch prior to `CMake: Configure` or `CMake: Build`.

When I manually trigger the python script, it only modifies the environment of the `os.system()`'s child process, not affecting the outside environment.  A way to confirm this is to add the line `os.system('set | tee /tmp/env')` after the `os.system('activate freecad')` call.  In the file `/tmp/env`, the variable `CONDA_PYTHON_EXE` should point to the `freecad` environment's installed python.  When I evaluate it, it still points to the base environment `python`.

The VS Code `cmake-kits.json` file has an [`environmentSetupScript`](https://github.com/microsoft/vscode-cmake-tools/blob/main/docs/kits.md#general-options) option which is executed by `call` on Windows and `source` on Linux and macOS.  This means the environment changes are performed in the current shell instance, which is used by CMake to inherit the environment (note: use "source activate freecad" not "activate freecad").

Reply:
Right so:
- the python script cannot change the parent environment
- even if it did the tasks run in different shells so it wouldn't work

I think this would be the only solution
```
    {
      "type": "shell",
      "label": "CMake: configure",
      "command": "${workspaceFolder}/.vscode/ActivateCondaIfUsingConda.sh ${command:cmake.activeConfigurePresetName}; cmake ${workspaceFolder} --preset ${command:cmake.activeConfigurePresetName}",
      "problemMatcher": [],
      "detail": "Configure the project"
    },
    {
      "type": "shell",
      "label": "CMake: build",
      "command": "${workspaceFolder}/.vscode/ActivateCondaIfUsingConda.sh ${command:cmake.activeConfigurePresetName}; cmake --build ${command:cmake.buildDirectory} --parallel",
      "group": {
        "kind": "build",
        "isDefault": true
      },
      "problemMatcher": [],
      "detail": "Default build task"
    }
```

This changes the tasks from a cmake task to a shell command. Then in one line we first set the environment and then manually call cmake. Note that the extension now does not call cmake internally anymore, we basically just use it for variable substitution. 
- Would something like this work?
- how to make the env setup script crossplatform?
- do we need to add a test task as well to activate the environment before testing?
- if it works can we remove the following lines from the cmake conda preset?
```
"cmakeExecutable": "${workspaceFolder}/.conda/freecad/bin/cmake",
"CMAKE_C_COMPILER": "${workspaceFolder}/.conda/freecad/bin/cc",
"CMAKE_CXX_COMPILER": "${workspaceFolder}/.conda/freecad/bin/c++"
```

Another question: what variable does `activate freecad` change exactly? could we simply put those environment variables in the conda preset?

Reply:
Prior to the MR I submitted, I spent about 4 weeks in getting the environment working on macOS and implemented in VS Code.  I had not planned for this effort, and mistakenly believed that compiling and debugging FreeCAD in VS Code on macOS would be straightforward as it is on Linux.  The FreeCAD development environment is simply a first task on a longer, self-funded project I am working on.

The instructions on the Wiki are no longer valid for macOS, and over time will not be valid for the other platforms either.  The reason is that the underlying dependencies are not fixed, and do not come from reproducible build environments.  `conda`'s purpose is to ensure a reproducible build and run environment for Python-based systems (such as FreeCAD).  Consequently, years ago @looooo set up the conda build and release system for the weekly builds and official releases.  This system has been very reliable in ensuring successful builds across platforms and environments.  The solution for macOS development, which works equally well on Linux, was to bring over the conda configuration used in the official releases and establish a VS Code environment using it.

During this effort, the only effective means of setting the environment I found was using the `environmentSetupScript` option in `cmake-kits.json` configuration provided by the `vscode-cmake-tools` extension, which has been added for exactly this purpose.

The `cmake-kits.json` is intended to provide a project to customize the discovery and configuration of toolchains, including per-platform configurations.  I do believe that this is the correct location for setting up the platform-specific configuration.  Previously the `cmake-kit` options were for `FreeCAD Linux` and `FreeCAD macOS` which would call the conda activation environment shell script and set the `CC` and `CXX` compiler paths.

What I did not succeed at was getting a manual configuration and build on Windows, and therefore did not port over the environment activation script.  A `FreeCAD Windows` would have been added, had I been successful.  I only spent a day on this task as I do not develop on, nor use Windows and the effort felt less than beneficial for me, especially since I was already 4 weeks behind on my project schedule and budget.

Reply:
Thanks for the detailed explanation. I want to reassure you I am not trying to remove conda, I am simply trying to configure vscode as best as possible and provide a configuration for all platforms. This includes making conda one of the options. On the other hand using the system provided libraries should be one of the options as well.

In an effort to make the build process more configurable I opted for CMakePresets, which from what I understand replaces cmake-kits. It is not possible to use both together.

I've tried the approach suggested in my last comment. From what I can see, at least on linux, it succeeds setting environment variables correctly before running cmake. I believe this would work on MacOS too.

Thanks for your patience and for helping sort this out.

Reply:
The tasks `CMake: configure` and `CMake: build` do not override the commands `CMake: configure` and `CMake: build`.  Consequently, running `CMake: configure` and `CMake: build` as normal did not run the script.

Reply:
I am aware, unfortunately I found out the hard way. Indeed you need to run the vscode task and not the extension command.